ding/policy/bc.py

```diff
@@ -0,0 +1,192 @@
+        if self._cuda:
+            data = to_device(data, self._device)
+        self._learn_model.train()
+        with self._timer:
+            if self.cfg.eval.evaluator.cfg_type == 'MetricSerialEvaluatorDict':
+                obs, action = data
+            else:
+                obs, action = data['obs'], data['action']
+            a_logit = self._learn_model.forward(obs)
+            loss = self._ce_loss(a_logit['logit'], action)
+        forward_time = self._timer.value
+        with self._timer:
+            self._optimizer.zero_grad()
+            loss.backward()
+        backward_time = self._timer.value
+        with self._timer:
+            if self._cfg.learn.multi_gpu:
+                self.sync_gradients(self._learn_model)
+        sync_time = self._timer.value
+        self._optimizer.step()
+        cur_lr = [param_group['lr'] for param_group in self._optimizer.param_groups]
+        cur_lr = sum(cur_lr) / len(cur_lr)
+        return {
+            'cur_lr': cur_lr,
+            'total_loss': loss.item(),
+            'forward_time': forward_time,
+            'backward_time': backward_time,
+            'sync_time': sync_time,
+        }
+
+    def _monitor_vars_learn(self):
+        return ['cur_lr', 'total_loss', 'forward_time', 'backward_time', 'sync_time']
+
+    def _init_eval(self):
+        self._eval_model = model_wrap(self._model, wrapper_name='argmax_sample')
+        self._eval_model.reset()
+
+    def _forward_eval(self, data):
+        if self.cfg.eval.evaluator.cfg_type == 'MetricSerialEvaluatorDict':
+            data = default_collate(list(data))
+            if self._cuda:
+                data = to_device(data, self._device)
+            self._eval_model.eval()
+            with torch.no_grad():
+                output = self._eval_model.forward(data)
+            if self._cuda:
+                output = to_device(output, 'cpu')
+            return output
+        else:
+            data_id = list(data.keys())
+            data = default_collate(list(data.values()))
+            if self._cuda:
+                data = to_device(data, self._device)
+            self._eval_model.eval()
+            with torch.no_grad():
+                output = self._eval_model.forward(data)
+            if self._cuda:
+                output = to_device(output, 'cpu')
+            output = default_decollate(output)
+            return {i: d for i, d in zip(data_id, output)}
+
+    def _init_collect(self) -> None:
+        r"""
+        Overview:
+            Collect mode init method. Called by ``self.__init__``.
+            Init traj and unroll length, collect model.
+            Enable the eps_greedy_sample
+        """
+        self._unroll_len = self._cfg.collect.unroll_len
+        #self._gamma = self._cfg.discount_factor  # necessary for parallel
+        #self._nstep = self._cfg.nstep  # necessary for parallel
+        self._collect_model = model_wrap(self._model, wrapper_name='eps_greedy_sample')
```

Comment:
why bc needs `collect` part

Reply:
In the latter part, DREX will inject noise into the BC policy (using epsilon greedy model wrapper). So I use the collect mode to do that.

Reply:
add some comments to explain this